### PR TITLE
Implement optimal message selection

### LINKED ISSUE #1054: Implement a correct chain abstraction for optimal message selection
**Task summary**

This was a result of investigating the failing tests in #1023 (PR: ##1037)
We need to re-write the MsgChain and MsgChainNode structures to mimic closely how lotus uses it.
Lotus's message pool implementation uses a array of msgChain nodes which provides both array and list semantics.

It's not entirely clear if it's possible to approach it 1:1 because of Rust's borrow check rules. What we have been exploring is the slotmap crate.

More details to follow soon on this issue.

**Specification reference**

- 

**Other information and links**

### PR DESCRIPTION
**Pull request by creativcoder**
**Summary of changes**
Changes introduced in this pull request:
- Closes #1054 and #1023 

**Reference issue to close (if applicable)**

Closes 

**Other information and links**

### CODE REVIEW

File: blockchain/message_pool/src/msgpool/selection.rs [hunk 1]
```diff
@@ -54,26 +72,374 @@ where
...
+            .read()
+            .await
+            .chain_compute_base_fee(&target_tipset)?;
+
+        // 0. Load messages from the target tipset; if it is the same as the current tipset in
+        //    the mpool, then this is just the pending messages
+        let mut pending = self.get_pending_messages(&cur_ts, &target_tipset).await?;
+
+        if pending.is_empty() {
+            // FIXME: lotus returns a nil, shouldn't we return an Err?
```
Comment (creativcoder) [inline blockchain/message_pool/src/msgpool/selection.rs:111]: @ec2 should we?

Comment (ec2) [reply blockchain/message_pool/src/msgpool/selection.rs:111]: No. If theres no pending messages, then we select no messages. Makes sense to return empty vec.

**Review by cryptoquick — COMMENTED** [review]
slotmap looks really neat! However, I'm afraid I'm not able to provide review without further context.

File: blockchain/message_pool/src/msg_chain.rs [hunk 2]
```diff
@@ -19,23 +260,50 @@ pub(crate) struct MsgChainNode {
...
+                    .cmp(&other.gas_reward)
+                    .eq(&Ordering::Greater)
+        {
+            return Ordering::Greater;
+        }
+
+        Ordering::Less
+    }
+
+    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (creativcoder) [reply blockchain/message_pool/src/msg_chain.rs:282]: Right, that's gone

File: blockchain/message_pool/src/msg_chain.rs [hunk 3]
```diff
@@ -1,15 +1,256 @@
...
+}
+
+/// Chains is an abstraction of a list of message chain nodes.
+/// It wraps a slotmap instance. key_vec is an additional requirement in order to satisfy
+/// optimal msg selection use cases, such as iteration in insertion order.
+/// The slotamap serves as a lookup table for nodes to get around the borrow checker rules.
+/// Each MsgChainNode contains only pointers as `NodeKey` to the entries in the map
+/// With this design, we get around the borrow checker rule issues when
+/// implementing the optimal selection algorithm.
+pub struct Chains {
```
[the remark this answers is not in the document]
Comment (creativcoder) [reply blockchain/message_pool/src/msg_chain.rs:31]: done

File: blockchain/message_pool/src/msgpool/selection.rs (hunk 1 above)
Comment (creativcoder) [reply blockchain/message_pool/src/msgpool/selection.rs:111]: Alright!

File: Cargo.lock [hunk 4]
```diff
@@ -3590,13 +3590,13 @@ checksum = "3576a87f2ba00f6f106fdfcd16db1d698d648a26ad8e0573cad8537c3c362d2a"
 
 [[package]]
 name = "lexical-core"
-version = "0.7.5"
+version = "0.7.4"
```
Comment (ec2) [inline Cargo.lock:3593]: run `cargo update -p lexical-core:0.7.4` and this should fix the nightly build check.

**Review by ec2 — APPROVED** [review]
LGTM aside from the clippy issues and nightly build check.

**Review by cryptoquick — COMMENTED** [review]
Thank you for the HackMD primer you made on [Optimal Message Selection](https://hackmd.io/@tbdrqGmwSXiPjxgteK3hMg/rkJmTfrvu#). It was very useful to understand the context of this PR.

I don't have specific comments on this PR, but I would like to ask a few general questions.

Do you have an idea of how many messages will be in the SlotMap each time?

Because the SlotMap is frequently cleared, I wonder if we shouldn't use the HopSlotMap, with a predefined (rough estimate) of capacity, using `with_capacity_and_key`, and then using the `clear` method to explicitly clear its contents.

Also, I noticed you removed the comment "TODO: Implement guess gas". Is that implemented now, say, via chain message? If so, are we checking against the new `MIN_GAS` constant anywhere the new, dynamic `min_gas` value should be used instead?

Finally, do you have any test cases? How have you observed that this new message selection algorithm works?

**Comment by creativcoder** [comment]
> Thank you for the HackMD primer you made on [Optimal Message Selection](https://hackmd.io/@tbdrqGmwSXiPjxgteK3hMg/rkJmTfrvu#). It was very useful to understand the context of this PR.
> 
> I don't have specific comments on this PR, but I would like to ask a few general questions.
> 
> Do you have an idea of how many messages will be in the SlotMap each time?

Glad you found it helpful Hunter. Um, @ec2 might have a better answer on that. When I ran the tests for optimal selection (3 of them) I get about 1437 messages picked out of the selection method. It will usually be variable though.

> 
> Because the SlotMap is frequently cleared, I wonder if we shouldn't use the HopSlotMap, with a predefined (rough estimate) of capacity, using `with_capacity_and_key`, and then using the `clear` method to explicitly clear its contents.

We don't need to delete anything from the map anytime during the algorithm, it gets de-allocated at end of the method. We just remove, invalidate pointers to them.
> 
> Also, I noticed you removed the comment "TODO: Implement guess gas". Is that implemented now, say, via chain message? If so, are we checking against the new `MIN_GAS` constant anywhere the new, dynamic `min_gas` value should be used instead?

So gas_guess is actually referring to a module from [lotus impl](https://github.com/filecoin-project/lotus/blob/master/chain/messagepool/gasguess/guessgas.go). As of now we only use two constants from it and we don't have the rest of entities being used anywhere in forest. Yeah, it's better to put the TODO comment back, actually at the top where MIN_GAS constant is defined.

> 
> Finally, do you have any test cases? How have you observed that this new message selection algorithm works?

Yes, `selection.rs` contains 3 test cases for optimal selection. Starts here: [selection.rs](https://github.com/ChainSafe/forest/pull/1086/files#diff-a783b11c9b7dbd811a7cbf71bc863d5bf68a34cfb6768c0c513dc82484b766b4R1020) They are direct ports of lotus' [selection_test.go](https://github.com/filecoin-project/lotus/blob/master/chain/messagepool/selection_test.go)

**Review by cryptoquick — APPROVED** [review]
I feel comfortable with those answers! Fantastic work!

I think we should get this in before we merge #1033 due to our changes to how the keystore is created (Memory, Persistent, Encrypted). We'll handle those changes when we update our branch.

**Comment by cryptoquick** [comment]
Also, don't forget to make the linter happy!

https://app.circleci.com/pipelines/github/ChainSafe/forest/4132/workflows/6168d5f9-bf19-40ad-a466-55f4d693c3c4/jobs/15493/parallel-runs/0/steps/0-111

File: Cargo.lock (hunk 4 above)
Comment (creativcoder) [reply Cargo.lock:3593]: I cannot update to 0.7.4, it always updates to 0.7.6, also crates.io page lists 0.7.4 in gray: https://crates.io/crates/lexical-core/versions Can we go with 0.7.6?

Comment (ec2) [reply Cargo.lock]: Well the issue is that you're on 0.7.4 right now. `cargo update -p lexical-core:0.7.4` upgrades it to the newest version according to semvar. So yeah, the point of that command IS to update to 0.7.6.

Comment (creativcoder) [reply Cargo.lock]: Ah I see. It's fixed.

**Pull request merged**